# Практическое задание №5 по теме "Сингулярное разложение матриц".

In [1]:
import numpy as np
np.set_printoptions(precision=3, suppress=True)

## Задание 1

Найдите посредством NumPy SVD для матрицы:

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$

___Решение___

In [2]:
A = np.array([[1,2,0],
              [0,0,5],
              [3,-4,2],
              [1,6,5],
              [0,1,0]])
print(f'Матрица A:\n{A}')

Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [3]:
# SVD
U, s, W = np.linalg.svd(A)
# Транспонируем матрицу W
V = W.T
# Создаем D из одномерного массива s
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [4]:
print(f'Матрица U:\n{U}')
print(f'Матрица D:\n{D}')
print(f'Матрица V:\n{V}')

Матрица U:
[[ 0.171  0.157 -0.531 -0.799 -0.162]
 [ 0.393 -0.529  0.613 -0.434  0.031]
 [-0.144 -0.824 -0.524  0.14   0.074]
 [ 0.888  0.061 -0.247  0.378 -0.06 ]
 [ 0.081  0.108 -0.082 -0.105  0.982]]
Матрица D:
[[8.825 0.    0.   ]
 [0.    6.141 0.   ]
 [0.    0.    2.533]
 [0.    0.    0.   ]
 [0.    0.    0.   ]]
Матрица V:
[[ 0.071 -0.367 -0.927]
 [ 0.717  0.665 -0.208]
 [ 0.693 -0.65   0.311]]


Проверка:

In [5]:
print(np.dot(np.dot(U, D), V.T))

[[ 1.  2. -0.]
 [-0.  0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [ 0.  1.  0.]]


___Ответ:___

$$U=\begin{pmatrix}
0.171 & 0.157 & -0.531 & -0.799 & -0.162\\
0.393 & -0.529 &  0.613 & -0.434 &  0.031\\
-0.144 & -0.824 & -0.524 &  0.14 &   0.074\\
0.888 &  0.061 & -0.247 &  0.378 & -0.06 \\
0.081 &  0.108 & -0.082 & -0.105 &  0.982
\end{pmatrix}$$

$$D=\begin{pmatrix}
8.825 & 0 & 0 \\
0 & 6.141 0 \\
0 & 0 & 2.533 \\
0 & 0 & 0 \\
0 & 0 & 0 
\end{pmatrix}$$

$$V=\begin{pmatrix}
0.071 & -0.367 & -0.927\\
0.717 & 0.665 & -0.208\\
0.693 & -0.65 & 0.311
\end{pmatrix}$$

---

## Задание 2

Для матрицы из предыдущего задания найдите:

    а) евклидову норму;
    
    б) норму Фробениуса.

___Решение___

__а)__

_Евклидова норма_:
    
$$\left \| A \right \|_{E}=\text{max}\left (\frac{\left \| Ax \right \|}{\left \| x \right \|}\right )$$ 

Евклидова норма матрицы равна евклидовой норме диагональной матрицы из её сингулярных чисел $D$. Максимальное значение полученного отношения будет равно максимальному сингулярному числу $\mu_{max}$, и, принимая во внимание факт сортировки по убыванию сингулярных чисел, получим:

$$\left \| A \right \|_{E}=\mu_{1}=8.825$$

__б)__
    
_Норма Фробениуса_:

$$\left \| A \right \|_{F}=\sqrt{\sum_{i=1}^{m}\sum_{j=1}^{n}a_{ij}^{2}}.$$

Когда известно сингулярное разложение матрицы, её норма Фробениуса вычисляется как:

$$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}=\sqrt{\sum_{k=1}^{3}\mu_{k}^{2}} =\sqrt{\mu_{1}^{2} + \mu_{2}^{2} + \mu_{3}^{2}} =\sqrt{8.825^{2} + 6.141^{2} + 2.533^{2}}= 11.04536$$

In [6]:
print(f"A_f = {np.round(np.sqrt(np.sum(np.power(s, 2))), 5)}")

A_f = 11.04536


___Ответ:___ ___а)___ $\left \| A \right \|_{E}=8.825$, $\;$ ___б)___ $\left \| A \right \|_{F}=11.04536$

---